File: rust/src/35.search-insert-position.rs
use crate::Solution;

impl Solution {
    pub fn search_insert(nums: Vec<i32>, target: i32) -> i32 {
        if target <= nums[0] {
            return 0;
        }

        let mut ok = 0_i32;
        let mut ng = nums.len() as i32;

        while ng - ok > 1 {
            let mid = (ok + ng) / 2;
            let num = nums[mid as usize];

            if num == target {
                return mid;
            } else if num < target {
                ok = mid;
            } else {
                ng = mid
            }
        }

        if nums[ok as usize] == target {
            ok
        } else {
            ng
        }
    }
}

File: ruby/35.search-insert-position.rb
# frozen_string_literal: true

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def search_insert(nums, target)
  return 0 if target <= nums[0]

  left = 0
  right = nums.size

  while right - left > 1
    mid = left + (right - left) / 2

    if target == nums[mid]
      return mid
    elsif target < nums[mid]
      right = mid
    else
      left = mid
    end
  end

  nums[left] == target ? left : right
end
